src/odemis/gui/cont/temperature.py

```diff
@@ -0,0 +1,124 @@
+Software Foundation.
+
+Odemis is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+Odemis. If not, see http://www.gnu.org/licenses/.
+
+"""
+
+import wx
+import collections
+import numpy
+import time
+import logging
+import odemis.gui as gui
+from odemis.gui import comp
+import odemis.model as md
+
+
+class TemperatureController(object):
+    '''
+    Temperature Controller
+    Displays the temperature from a thermostat component in the main GUI on a predefined control
+    and updates at a set sampling period, which is polled.
+    '''
+
+    def __init__(self, main_frame, thermostat, max_duration=5):
+        '''
+        main_frame: (wx) main xrc frame of the GUI
+        thermostat: (HwComponent) a component with the VA "temperature"
+        max_duration: (float) in seconds, the maximum duration
+        '''
+
+        if not hasattr(thermostat, "temperature"):
+            raise ValueError("Thermostat is missing the temperature VA")
+        self.thermostat = thermostat
+
+        self.temperature_label = main_frame.temperature_display
+        self._fg_color = self.temperature_label.GetForegroundColour()
+        self.temperature_label.Show()
+
+        self._time_last_warning = 0  # time of the last warning
+        self._time_last_speed_warning = 0
+
+        # temperature history is a deque of tuples (timestamp, temperature)
+        self.temperature_history = collections.deque([(time.time(), self.thermostat.temperature.value)])
+        self.max_duration = max_duration
+        self.thermostat.temperature.subscribe(self._updateTemperature)
+
+    def _updateTemperature(self, temperature):
+        """
+        Subscriber to the temperature VA
+        """
+        # add the new temperature to the history with a timestamp
+        self.temperature_history.append((time.time(), temperature))
+        wx.CallAfter(self.temperature_label.SetLabel, u"{temp:.2f}°C".format(temp=temperature))
+        
+        # clear out values of the history that are too old i.e. older than the max_duration
+        while len(self.temperature_history) > 0:
+            t, _ = self.temperature_history[0]
+            if t < time.time() - self.max_duration:
+                self.temperature_history.popleft()
+            else:
+                break
+
+        # Get boundaries
+        safe_range = self.thermostat.getMetadata().get(md.MD_FAV_POS_RANGE, None)
+        safe_speed = self.thermostat.getMetadata().get(md.MD_SAFE_SPEED_RANGE, None)
+
+        # Calculate a linear regression of the temperature history to determine the speed
+        l = list(zip(*self.temperature_history))  # convert [(time, temp), ...] to [[time, ...], [temp...]]
+        try:
+            speed, _ = numpy.polyfit(l[0], l[1], 1)
```

Comment:
This fails for simple cases where the len(l) < 2. It would be relatively easy to trigger if for some reason the temperature hasn't been updated in a while. Such case should *not* show up as an error in the GUI. So I'd suggest to explicitly check for that first, and return immediately in such case.